crates/chain/src/local_chain.rs

```diff
@@ -19,215 +170,250 @@ impl ChainOracle for LocalChain {
-    /// this will result in an [`InsertBlockNotMatchingError`].
-    pub fn insert_block(
-        &mut self,
-        block_id: BlockId,
-    ) -> Result<ChangeSet, InsertBlockNotMatchingError> {
-        let mut update = Self::from_blocks(self.tip());
-
-        if let Some(original_hash) = update.blocks.insert(block_id.height, block_id.hash) {
+    /// Replacing the block hash of an existing checkpoint will result in an error.
+    pub fn insert_block(&mut self, block_id: BlockId) -> Result<ChangeSet, InsertBlockError> {
+        if let Some(&original_hash) = self.index.get(&block_id.height) {
             if original_hash != block_id.hash {
-                return Err(InsertBlockNotMatchingError {
+                return Err(InsertBlockError {
                     height: block_id.height,
                     original_hash,
                     update_hash: block_id.hash,
                 });
+            } else {
+                return Ok(ChangeSet::default());
             }
         }
 
-        Ok(self.apply_update(update).expect("should always connect"))
+        let mut changeset = ChangeSet::default();
+        changeset.insert(block_id.height, Some(block_id.hash));
+        self.apply_changeset(&changeset);
+        Ok(changeset)
     }
-}
 
-/// This is the return value of [`determine_changeset`] and represents changes to [`LocalChain`].
-///
-/// [`determine_changeset`]: LocalChain::determine_changeset
-pub type ChangeSet = BTreeMap<u32, Option<BlockHash>>;
+    /// Reindex the heights in the chain from (and including) `from` height
+    fn reindex(&mut self, from: u32) {
+        let _ = self.index.split_off(&from);
+        for cp in self.iter_checkpoints() {
+            if cp.height() < from {
+                break;
+            }
+            self.index.insert(cp.height(), cp.hash());
+        }
+    }
 
-/// Represents an update failure of [`LocalChain`] due to the update not connecting to the original
-/// chain.
-///
-/// The update cannot be applied to the chain because the chain suffix it represents did not
-/// connect to the existing chain. This error case contains the checkpoint height to include so
-/// that the chains can connect.
-#[derive(Clone, Debug, PartialEq)]
-pub struct UpdateNotConnectedError(pub u32);
+    /// Derives an initial [`ChangeSet`], meaning that it can be applied to an empty chain to
+    /// recover the current chain.
+    pub fn initial_changeset(&self) -> ChangeSet {
+        self.index.iter().map(|(k, v)| (*k, Some(*v))).collect()
+    }
 
-impl core::fmt::Display for UpdateNotConnectedError {
-    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(
-            f,
-            "the update cannot connect with the chain, try include block at height {}",
-            self.0
-        )
+    /// Iterate over checkpoints in decending height order.
+    pub fn iter_checkpoints(&self) -> CheckPointIter {
```

Comment:
Do you think it makes sense to have the method name reflect the fact that the checkpoints are iterated backwards, from highest to lowest height?

Reply:
I think that will provide better clarity. Do you have any suggestions? @LLFourn what are your thoughts?

Reply:
Delete the method. You can just get the tip and iter backwards right?